core/src/appleMain/kotlin/CBPeripheralCoreBluetoothPeripheral.kt

```diff
@@ -393,9 +401,41 @@ internal class CBPeripheralCoreBluetoothPeripheral(
         }
 
         val platformDescriptor = discoveredServices.obtain(descriptor)
-        return connection.execute<DidUpdateValueForDescriptor> {
+        val updatedDescriptor = connection.execute<DidUpdateValueForDescriptor> {
             centralManager.read(cbPeripheral, platformDescriptor)
-        }.descriptor.value as NSData
+        }.descriptor
+
+        // CBDescriptor "value" property type depends on cbDescriptor.UUID.UUIDString
+        // https://developer.apple.com/documentation/corebluetooth/characteristic-descriptors
+        // see type conversion table https://github.com/JuulLabs/kable/pull/706
+        return when (val value = updatedDescriptor.value) {
+            is NSData -> value
+            is NSString -> value.dataUsingEncoding(NSUTF8StringEncoding) ?: byteArrayOf().toNSData().also {
+                logger.warn {
+                    message = "Failed to decode descriptor"
+                    detail("type", "NSString")
+                    detail(descriptor)
+                }
+            }
+            is NSNumber -> {
+                when (updatedDescriptor.UUID.UUIDString) {
+                    CBUUIDCharacteristicExtendedPropertiesString,
+                    CBUUIDClientCharacteristicConfigurationString,
+                    CBUUIDServerCharacteristicConfigurationString,
+                    -> value.unsignedShortValue.toNSData()
+                    else -> value.unsignedLongValue.toNSData()
+                }
+            }
+            is UInt16 -> value.toNSData()
```

Comment:
I'm guessing you have this here for `CBUUIDL2CAPPSMCharacteristicString`?
Although to documentation lists it as `UInt16`, I'd be surprised if this didn't come in as an `NSNumber`?
Aside from [CBUUIDL2CAPPSMCharacteristicString](https://developer.apple.com/documentation/corebluetooth/cbuuidl2cappsmcharacteristicstring), did you see any indication of what time this comes in as?

Reply:
Yes I've kept `Uint16` for `CBUUIDL2CAPPSMCharacteristicString`

> Aside from [CBUUIDL2CAPPSMCharacteristicString](https://developer.apple.com/documentation/corebluetooth/cbuuidl2cappsmcharacteristicstring), did you see any indication of what type this comes in as?

Only other reference was found by @davidtaylor-juul

>  Thanks for the detail. I agree with your deductions in the other comment as well. The only reference I found in the docs is for the type alias https://developer.apple.com/documentation/corebluetooth/cbl2cappsm/ which also hints that this the value for `CBUUIDL2CAPPSMCharacteristicString` is exposed as a raw 16bit integer, and is not a `NSNumber` or `NSData` object. It is for `L2CAP` so to test a real world example would require reading characteristics from a device that supports `L2CAP` channels. I'll update this thread if I find a concrete example.

and seems to indicate that it might return as `UInt16`, since `CBL2CAPPSM` is an alias for `UInt16`.

On the other hand, what you have found here (although it is for `CBCharacteristic` [value property](https://developer.apple.com/documentation/corebluetooth/cbcharacteristic/value) which is defined as `NSData` and not for `CBDescriptor` value property which is defined as `Any`)

> Not sure if it confirms your suspicions @davidtaylor-juul, but [this code](https://github.com/clostra/newnode/blob/6e64362fb2dd97cc58ee4b1624c466a85a43dc9f/ios/Bluetooth.m#L196-L231) retrieves the NSData as-is (without UInt16 casting, as far as I can tell, at least).

Seems to indicate it might be returned as `NSData`.

Given all these considerations, I think I can add management for `CBUUIDL2CAPPSMCharacteristicString` in the `NSNumber` block so that:

- if it is actually returned as `NSData` as indicated by the code you found for  `CBCharacteristic`  we are already covered by the current code. It will be returned directly
- if it is actually a `UInt16` as Apple doc hints, we are already covered by the current code. It will be converted and returned.
- if it is, for some reason,  actually returned as `NSNumber` I will add specific management and convert it to a two bytes array as documented by GATT Part G document.

I'll update the PR in a moment, let me know what you think about this.

Reply:
IMHO we could omit this case, but I suppose there is no harm leaving it in.